### PR TITLE
feat: product search

### LINKED ISSUE #105: [FEAT] Product Search Endpoint - BACKEND
## Description

This feature involves creating a product search endpoint for the backend system. The endpoint will allow users to search for products by name, category, and price range. Additional search features can be added to enhance the functionality.

## Acceptance Criteria

- Users can search for products by name.
- Users can search for products by category.
- Users can filter search results by price range.
- The endpoint should return accurate search results based on the given criteria.
- The search functionality should be efficient and performant, even with a large dataset.
- Proper error handling and validation should be implemented for search queries.

## Purpose

This feature aims to provide users with a robust and flexible product search capability, making it easier for them to find specific products based on various criteria. This will improve the overall user experience and increase the application's usability.

## Requirements	

- Implement a RESTful endpoint for product search.
- Use appropriate query parameters to handle search criteria (name, category, price range).
- Integrate with the existing product database to fetch and filter results.
- Ensure the endpoint adheres to best practices for API design and security.
- Write unit and integration tests to validate the functionality and performance of the endpoint.

### URL
```bash
/api/v1/products/search
```

### METHOD
```
GET
```
### Query Parameters
- `name (optional)`: The name of the product to search for.
- `category (optional)`: The category of the product to search for.
- `minPrice (optional)`: The minimum price of the product to search for.
- `maxPrice (optional)`: The maximum price of the product to search for.

### Example Request

```GET /api/v1/products/search?name=t-shirt&category=kids&minPrice=100&maxPrice=1000```

#### Success Response(json)
```json
{
  "success": true,
  "statusCode":200,
  "products": [
    {
      "id": 1,
      "name": "White T-shirt with image",
      "category": "kids",
      "price": 200,
      "description": "A white T-shirt with tom and jerry image ",
      "imageUrl": "http://example.com/images/white_t_shirt_tom_jerry_image.jpg"
    },
    {
      "id": 2,
      "name": "Black T-shirt",
      "category": "kids",
      "price": 150,
      "description": "A Black t-shirt",
      "imageUrl": "http://example.com/images/black_t_shirt.jpg"
    }
  ]
}

```
#### Empty Response(json)
```json
{
  "success": true,
  "products": [],
  "statusCode":204
}
```
#### Validation Error Response(json)
```json
{
  "success": false,
  "errors": [
    {
        "parameter":"name",
        "message":"name is  a required parameter"
    }
  ]
  "statusCode":422
}
```
#### Bad Request Response(json)
```json
{
  "success": false,
  "message": "Bad Request",
  "statusCode":400
}
```
#### Error Response(json)
```json
{
  "success": false,
  "message": "Server Error",
  "statusCode":500
}
```

## Task

- [ ] Design the product search endpoint.
- [ ] Implement query parameters for name, category, and price range.
- [ ] Integrate the search endpoint with the existing product database.
- [ ] Implement error handling and validation for search queries.
- [ ] Write unit tests for the search functionality.
- [ ] Write integration tests for the search endpoint.
- [ ] Document the API endpoint and usage.

## Expected Outcome

- A functional product search endpoint that meets all acceptance criteria.
- Users can efficiently search for products using various criteria.
- The search results are accurate and relevant to the user's query.
- The system handles a high volume of search requests without significant performance degradation.

### PR DESCRIPTION
**Pull request by GreyyDaze**
### **Description**

This PR closes #105 and implements api endpoitn for searching products. The endpoint `/api/v1/products/search` allows users to search for products based on name, category, and price range. The endpoint supports filtering by various criteria and provides accurate and efficient search results. The implementation includes error handling and validation to ensure robust functionality.

### **Testing Instructions**

1. **Open Postman.**
2. **Set the request type to GET.**
3. **Enter the API URL:**
   ```
   http://localhost:3000/api/v1/products/search
   ```
4. **Add Query Parameters:**
   - `name` (optional): Enter a product name to search for, e.g., `t-shirt`.
   - `category` (optional): Enter a product category to filter by, e.g., `kids`.
   - `minPrice` (optional): Enter the minimum price to filter by, e.g., `100`.
   - `maxPrice` (optional): Enter the maximum price to filter by, e.g., `1000`.
5. **Click Send.**
6. **Check the response:**

   - **Success Response with Products Found:**
     ```json
     {
       "success": true,
       "statusCode": 200,
       "products": [
         {
           "id": 1,
           "name": "White T-shirt with image",
           "category": "kids",
           "price": 200,
           "description": "A white T-shirt with Tom and Jerry image",
           "imageUrl": "http://example.com/images/white_t_shirt_tom_jerry_image.jpg"
         },
         {
           "id": 2,
           "name": "Black T-shirt",
           "category": "kids",
           "price": 150,
           "description": "A Black T-shirt",
           "imageUrl": "http://example.com/images/black_t_shirt.jpg"
         }
       ]
     }
     ```

  - **No Products Found Response:**
     ```json
     {
       "status": "No Content",
       "status_code": 204,
       "message": "No products found"
     }
     ```

   - **Bad Request Response:**
     ```json
     {
       "success": false,
       "message": "Bad Request",
       "statusCode": 400
     }
     ```

   - **Error Response:**
     ```json
     {
       "success": false,
       "message": "Server Error",
       "statusCode": 500
     }
     ```

### **Tasks Completed**

- [X] Implemented a GET endpoint `/api/v1/products/search` for product search functionality.
- [X] Added query parameters for searching by name, category, and filtering by price range.
- [X] Integrated with the existing product database to fetch and filter results.
- [X] Implemented error handling and validation for search queries.
- [X] Created unit tests for the search functionality.
- [X] Updated API documentation to reflect the new endpoint and usage.

### CODE REVIEW

File: src/modules/products/products.service.ts
Comment (Heba-WebDev) [inline src/modules/products/products.service.ts:1]: Please remove the console logs.

File: src/modules/products/tests/products.service.spec.ts
Comment (Heba-WebDev) [inline src/modules/products/tests/products.service.spec.ts:1]: Please remove the uncommented tests

Comment (GreyyDaze) [reply src/modules/products/tests/products.service.spec.ts:1]: fix

File: src/modules/products/products.service.ts
Comment (GreyyDaze) [reply src/modules/products/products.service.ts:1]: fix

**Review by Heba-WebDev — APPROVED** [review]

**Review by rugue — APPROVED** [review]

**Review by TheCodeGhinux — APPROVED** [review]

**Pull request merged**